osquery/distributed/plugins/bro_distributed.cpp

```diff
@@ -0,0 +1,336 @@
+    return s_mq_indi;
+  }
+
+  // Set Broker groups and subscribe to group topics
+  Status s_groups = parseBrokerGroups(FLAGS_bro_groups, startup_groups_);
+  if (!s_groups.ok()) {
+    return s_groups;
+  }
+  for (const auto& g : startup_groups_) {
+    bm.addGroup(g);
+  }
+
+  // Connect to Bro
+  auto status_broker = bm.peerEndpoint(FLAGS_bro_ip, FLAGS_bro_port);
+  if (!status_broker.ok()) {
+    return status_broker;
+  }
+  VLOG(1) << "Broker connection established";
+
+  // Send announce message
+  auto status_announce = bm.announce();
+  if (!status_announce.ok()) {
+    return status_announce;
+  }
+
+  return Status(0, "OK");
+}
+
+inline Status processMessage(const broker::message& msg,
+                             const std::string& topic,
+                             std::vector<DistributedQueryRequest>& oT_queries) {
+  BrokerManager& bm = BrokerManager::get();
+  QueryManager& qm = QueryManager::get();
+
+  // Check Event Type
+  if (msg.size() < 1 or !broker::is<std::string>(msg[0])) {
+    return Status(3, "No or invalid event name when processing message");
+  }
+  std::string eventName = *broker::get<std::string>(msg[0]);
+  LOG(INFO) << "Received event '" << eventName << "' on topic '" << topic
+            << "'";
+
+  // osquery::host_execute
+  if (eventName == bm.EVENT_HOST_EXECUTE) {
+    // One-Time Query Execution
+    SubscriptionRequest sr;
+    createSubscriptionRequest("EXECUTE", msg, topic, sr);
+    std::string newQID = qm.addOneTimeQueryEntry(sr);
+    if (newQID.empty()) {
+      return Status(8, "Unable to add Broker Query Entry");
+    }
+    DistributedQueryRequest dqr;
+    dqr.id = newQID;
+    dqr.query = sr.query;
+    oT_queries.push_back(dqr);
+    return Status(0, "OK");
+
+    // osquery::host_join
+  } else if (eventName == bm.EVENT_HOST_JOIN) {
+    std::string newGroup = *broker::get<std::string>(msg[1]);
+    return bm.addGroup(newGroup);
+
+    // osquery::host_leave
+  } else if (eventName == bm.EVENT_HOST_LEAVE) {
+    std::string newGroup = *broker::get<std::string>(msg[1]);
+    return bm.removeGroup(newGroup);
+
+    // osquery::host_subscribe
+  } else if (eventName == bm.EVENT_HOST_SUBSCRIBE) {
+    // New SQL Query Request
+    SubscriptionRequest sr;
+    createSubscriptionRequest("SUBSCRIBE", msg, topic, sr);
+    Status s_sub = qm.addScheduleQueryEntry(sr);
+    if (!s_sub.ok()) {
+      return s_sub;
+    }
+
+    // osquery::host_unsubscribe
+  } else if (eventName == bm.EVENT_HOST_UNSUBSCRIBE) {
+    // SQL Query Cancel
+    SubscriptionRequest sr;
+    createSubscriptionRequest("UNSUBSCRIBE", msg, topic, sr);
+    // TODO: find an UNIQUE identifier (currently the exact sql string)
+    std::string query = sr.query;
+
+    Status s_unsub = qm.removeQueryEntry(query);
+    if (!s_unsub.ok()) {
+      return s_unsub;
+    }
+
+  } else if (eventName == "osquery::host_test") {
+  } else {
+    // Unkown Message
+    return Status(7, "Unknown event name '" + eventName + "'");
```

Comment:
What is the projects practice on this? I dont care to use 0/1 only. I introduced different error numbers to help in debugging and finding the root cause if a functions can potentially fail for different reasons. These codes are therefore not global but function-specific.
Enum about Bro Statuses might be overkilled, I will limit all error codes to 1.